File: erpnext/production/report/production_target_report/production_target_report.py
# Copyright (c) 2013, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from erpnext.accounts.utils import get_child_cost_centers, get_period_date


def execute(filters=None):
       
	columns = get_columns(filters)
        data = get_data(filters)

        return columns, data

def get_columns(filters):
	if filters.uinput == "Production":
        	cols= [
             		("Production Group") + ":Data:140",
                	("Target Qty") + ":Data:90"
        	]
	else:
		cols=[
			("Disposal Group") + ":Data:140",
			("Disposal Qty") + ":Data:97"
		]
	return cols 

def get_data(filters):
	all_ccs = get_child_cost_centers(filters.cost_center)

	if not filters.uinput:
		return []
	query = "select pdi.production_group, sum(pdi.quantity) from `tab{0} Target Item` pdi, `tabProduction Target` pt where pt.name = pdi.parent".format(filters.uinput) 

	if filters.get("branch"):
		query += " and pt.cost_center = \'" + str(filters.cost_center) + "\'"	
	else:	
		query += " and pt.cost_center in {0} ".format(tuple(all_ccs))
			
	if filters.get("location"):
		query += " and pt.location = \'" + str(filters.location) + "\'"
	if filters.get("fiscal_year"):
		query += " and pt.fiscal_year = \'" + str(filters.fiscal_year) + "\'"	
	query += " group by pdi.production_group " 

	return frappe.db.sql(query)

File: erpnext/production/report/production_target_report/production_target_report.js
// Copyright (c) 2016, Frappe Technologies Pvt. Ltd. and contributors
// For license information, please see license.txt

frappe.query_reports["Production Target Report"] = {
	"filters": [
		{
                        "fieldname": "company",
                        "label": ("Company"),
                        "fieldtype": "Link",
                        "options": "Company",
                        "default": frappe.defaults.get_user_default("Company"),
                        "reqd": 1
                },
		{
			"fieldname": "fiscal_year",
			"label": __("Fiscal Year"),
			"fieldtype": "Link",
			"options": "Fiscal Year",
			"reqd": 1,
		}, 
		{
                        "fieldname": "uinput",
                        "label": ("Options"),
                        "fieldtype": "Select",
                        "options": ["Production", "Disposal"],
                        "reqd" : 1,
                },

		{
                        "fieldname": "cost_center",
                        "label": ("Parent Branch"),
                        "fieldtype": "Link",
                        "options": "Cost Center",
                        "get_query": function() {
                                var company = frappe.query_report.filters_by_name.company.get_value();
                                return {
                                        'doctype': "Cost Center",
                                        'filters': [
                                                ['is_disabled', '!=', '1'],
                                                ['company', '=', company]
                                        ]
                                }
                        },
                        "on_change": function(query_report) {
                                var cost_center = query_report.get_values().cost_center;
                                //query_report.filters_by_name.branch.set_input(null);
                                //query_report.filters_by_name.location.set_input(null);
                                //query_report.filters_by_name.adhoc_production.set_input(null);
                                query_report.trigger_refresh();
                                if (!cost_center) {
                                        return;
                                }
                                frappe.call({
                                        method: "erpnext.custom_utils.get_branch_from_cost_center",
                                        args: {
                                                "cost_center": cost_center,
                                        },
                                        callback: function(r) {
                                                query_report.filters_by_name.branch.set_input(r.message)
                                                query_report.trigger_refresh();
                                        }
                                })
                        },
                        "reqd": 1,
                 },
		 {
                        "fieldname": "branch",
                        "label": ("Branch"),
                        "fieldtype": "Link",
                        "options": "Branch",
                        "read_only": 1,
                        "get_query": function() {
                                var company = frappe.query_report.filters_by_name.company.get_value();
                                return {"doctype": "Branch", "filters": {"company": company, "is_disabled": 0}}
                        }
		
                },
               	{
                        "fieldname": "location",
                        "label": ("Location"),
                        "fieldtype": "Link",
                        "options": "Location",
                        "get_query": function() {
                                var branch = frappe.query_report.filters_by_name.branch.get_value();
                                return {"doctype": "Location", "filters": {"branch": branch, "is_disabled": 0}}
                        }
                }
	]
}
